fedcloud_monitoring_tools/vo_mon.py

```diff
@@ -0,0 +1,157 @@
+"""VO-level testing for a given site in a VO"""
+
+import io
+import os
+import time
+from datetime import datetime, timezone
+
+import click
+import paramiko
+from fabric import Connection
+from imclient import IMClient
+
+IM_REST_API = "https://im.egi.eu/im"
+AUTH_FILE = "auth.dat"
+MAX_NUM_CHECKS = 10
+
+
+class VOTestException(Exception):
+    pass
+
+
+class VOTest:
+    """Helper class to call im-client easily"""
+
+    def __init__(self, vo, site, token):
+        self.vo = vo
+        self.site = site
+        self.token = token
+        self.now = datetime.now(timezone.utc)
+
+    def create_auth_file(self, filepath):
+        with open(filepath, "w") as output_file:
+            output_file.write(
+                "id = im; type = InfrastructureManager; token = {}\n".format(self.token)
+            )
+            output_file.write(
+                "id = egi; type = EGI; host = {}; vo = {}; token = {}\n".format(
+                    self.site, self.vo, self.token
+                )
+            )
+
+    def delete_auth_file(self, filepath):
+        if os.path.exists(filepath):
+            os.remove(filepath)
+
+    def create_vm_tosca_template(self):
+        inf_desc = """
+            tosca_definitions_version: tosca_simple_yaml_1_0
+            imports:
+            - grycap_custom_types: https://raw.githubusercontent.com/grycap/tosca/main/custom_types.yaml
+            topology_template:
+              node_templates:
+                simple_node:
+                  type: tosca.nodes.indigo.Compute
+                  capabilities:
+                    endpoint:
+                      properties:
+                        network_name: PUBLIC
+                    host:
+                      properties:
+                        num_cpus: 2
+                        mem_size: 4 GB
+                    os:
+                      properties:
+                        image: appdb://SITE/egi.ubuntu.24.04?VO
+              outputs:
+                node_ip:
+                  value: { get_attribute: [ simple_node, public_address, 0 ] }
+                node_creds:
+                  value: { get_attribute: [ simple_node, endpoint, credential, 0 ] }
+        """
+        return inf_desc.replace("SITE", self.site).replace("VO", self.vo)
```

Comment:
I would use instead of replace something like `"{site} {vo}".format(site=self.site, vo=self.vo)`

Reply:
ok, you probably don't want to start escaping every `{`. keep it as it is